File: src/flag_gems/ops/embedding.py
import logging

import torch
import triton
import triton.language as tl

import paddle
from paddle.autograd import PyLayer
from flag_gems.runtime import torch_device_fn
from flag_gems.utils import libentry
from flag_gems.utils import triton_lang_extension as tle

logger = logging.getLogger(__name__)


@libentry()
@triton.jit
def embedding_kernel(
    out_ptr,  # pointer to the output
    in_ptr,  # pointer to the input
    weight_ptr,  # pointer to the weights
    N: tl.constexpr,  # number of columns in X
    BLOCK_SIZE: tl.constexpr,
):
    pid = tle.program_id(0)
    out_ptr += pid * N
    in_ptr += pid

    mask = tl.arange(0, BLOCK_SIZE) < N
    cols = tl.arange(0, BLOCK_SIZE)

    row_idx = tl.load(in_ptr)
    weight_ptr += row_idx.to(tl.int32) * N
    embedding_weight = tl.load(weight_ptr + cols, mask, other=0.0)
    tl.store(out_ptr + cols, embedding_weight, mask)


@libentry()
@triton.jit
def indice_freq_kernel(
    indices_freq,
    indices,  # pointer to the input
    elem_cnt: tl.constexpr,  # number of columns in X
    INDICE_BLOCK_SIZE: tl.constexpr,
):
    pid = tle.program_id(0)
    block_start = pid * INDICE_BLOCK_SIZE

    offsets = block_start + tl.arange(0, INDICE_BLOCK_SIZE)
    mask = offsets < elem_cnt

    index_element = tl.load(indices + offsets, mask=mask)
    tl.atomic_add(indices_freq + index_element, 1, mask=mask)


@libentry()
@triton.jit(do_not_specialize=["padding_idx"])
def embedding_backward_kernel(
    grad_in,  # pointer to the gradient input
    grad_out,  # pointer to the gradient output
    indices,  # pointer to the input
    padding_idx,  # padding_idx
    HAS_PADDING_IDX: tl.constexpr,
    N: tl.constexpr,  # number of columns in X
    BLOCK_SIZE: tl.constexpr,
):
    pid = tle.program_id(0)
    grad_out += pid * N
    indices += pid

    mask = tl.arange(0, BLOCK_SIZE) < N
    cols = tl.arange(0, BLOCK_SIZE)

    row_idx = tl.load(indices).to(tl.int32)
    if not HAS_PADDING_IDX:
        grad_in += row_idx * N
        embedding_grad = tl.load(grad_out + cols, mask, other=0.0)
        if tl.constexpr(embedding_grad.dtype.is_bf16()):
            embedding_grad = embedding_grad.to(tl.float32)
        tl.atomic_add(grad_in + cols, embedding_grad, mask=mask)
    else:
        if row_idx != padding_idx:
            grad_in += row_idx * N
            embedding_grad = tl.load(grad_out + cols, mask, other=0.0)
            if tl.constexpr(embedding_grad.dtype.is_bf16()):
                embedding_grad = embedding_grad.to(tl.float32)
            tl.atomic_add(grad_in + cols, embedding_grad, mask=mask)


@libentry()
@triton.jit(do_not_specialize=["n_rows"])
def embedding_grad_scale_kernel(
    grad_out,
    indice_freq,
    n_rows,
    N,
    BLOCK_SIZE: tl.constexpr,
):
    row_start = tle.program_id(0)
    row_step = tle.num_programs(0)

    for row_idx in range(row_start, n_rows, row_step):
        embedding_scale = 1.0
        indice_freq_val = tl.load(indice_freq + row_idx)
        if indice_freq_val > 1:
            embedding_scale = 1.0 / indice_freq_val

        cols = tl.arange(0, BLOCK_SIZE)
        mask = tl.arange(0, BLOCK_SIZE) < N
        embedding_grad = tl.load(grad_out + row_idx * N + cols, mask=mask)
        scaled_embedding_grad = embedding_grad * embedding_scale
        tl.store(grad_out + row_idx * N + cols, scaled_embedding_grad, mask=mask)


def embedding(indices, weight, padding_idx=-1, scale_grad_by_freq=False, sparse=False):
    print("前向")
    logger.debug("GEMS EMBEDDING FORWARD")
    assert not sparse, "Currently do not support sparse format"

    M = indices.numel().item()
    N = weight.shape[-1]

    BLOCK_SIZE = triton.next_power_of_2(N)
    # TODO: remove contiguous enforcement
    indices = indices.contiguous()
    weight = weight.contiguous()
    output = torch.empty((*indices.shape, N), device=indices.place, dtype=weight.dtype)

    with torch_device_fn.device(weight.place):
        embedding_kernel[M,](output, indices, weight, N, BLOCK_SIZE)

    return output


def embedding_backward(
    grad_outputs,
    indices,
    num_weights,
    padding_idx=-1,
    scale_grad_by_freq=False,
    sparse=False,
):
    logger.debug("GEMS EMBEDDING BACKWARD")
    assert not sparse, "Currently do not support sparse format"
    print("反向")

    M = indices.numel()
    N = grad_outputs.shape[-1]

    grad_inputs = torch.zeros(
        (num_weights, grad_outputs.shape[-1]),
        device=grad_outputs.place,
        dtype=(
            torch.float32
            if grad_outputs.dtype is torch.bfloat16
            else grad_outputs.dtype
        ),
    )

    if scale_grad_by_freq:
        indice_freq = torch.zeros(
            (num_weights,),
            requires_grad=False,
            device=grad_outputs.place,
            dtype=torch.int32,
        )
        INDICE_BLOCK_SIZE = 256
        indice_grid = (triton.cdiv(M, INDICE_BLOCK_SIZE),)

        with torch_device_fn.device(grad_outputs.place):
            indice_freq_kernel[indice_grid](indice_freq, indices, M, INDICE_BLOCK_SIZE)
    else:
        indice_freq = None

    BLOCK_SIZE = triton.next_power_of_2(N)

    HAS_PADDING_IDX = padding_idx is not None

    with torch_device_fn.device(grad_outputs.place):
        embedding_backward_kernel[M,](
            grad_inputs,
            grad_outputs,
            indices,
            padding_idx,
            HAS_PADDING_IDX,
            N,
            BLOCK_SIZE,
        )

    if scale_grad_by_freq:
        with torch_device_fn.device(grad_outputs.place):
            embedding_grad_scale_kernel[M,](
                grad_inputs, indice_freq, num_weights, N, BLOCK_SIZE
            )
    return (
        grad_inputs.to(torch.bfloat16)
        if grad_outputs.dtype is torch.bfloat16
        else grad_inputs
    )

class Embedding(PyLayer):
    @staticmethod
    def forward(ctx, indices, weight, padding_idx=-1, scale_grad_by_freq=False, sparse=False):
        # 保存反向传播需要的参数
        ctx.save_for_backward(indices, weight)
        ctx.padding_idx = padding_idx
        ctx.scale_grad_by_freq = scale_grad_by_freq
        ctx.sparse = sparse
        
        # 直接调用现有的前向函数
        return embedding(indices, weight, padding_idx, scale_grad_by_freq, sparse)
    
    @staticmethod
    def backward(ctx, grad_outputs):
        # 获取保存的参数
        indices, weight = ctx.saved_tensor()
        padding_idx = ctx.padding_idx
        scale_grad_by_freq = ctx.scale_grad_by_freq
        sparse = ctx.sparse
        
        # 直接调用现有的反向函数
        grad_weight = embedding_backward(
            grad_outputs, indices, weight.shape[0], 
            padding_idx, scale_grad_by_freq, sparse
        )
        
        # 返回梯度：indices, weight, padding_idx, scale_grad_by_freq, sparse
        return  None, grad_weight
embedding_paddle= Embedding.apply

File: lib/embedding.cpp
#include "flag_gems/operators.h"
#include "flag_gems/utils.h"

#include <iostream>
#include <tuple>
#include "c10/cuda/CUDAStream.h"
#include "triton_jit/triton_jit_function.h"

namespace flag_gems {
using namespace triton_jit;

at::Tensor embedding(const at::Tensor &weight,
                     const at::Tensor &indices,
                     int64_t padding_idx,
                     bool scale_grad_by_freq,
                     bool sparse) {
  TORCH_CHECK(!sparse, "Currently do not support sparse format");
  int64_t M = indices.numel();
  int64_t N = weight.size(-1);
  int64_t BLOCK_SIZE = utils::next_power_of_2(N);
  at::Tensor contiguous_indices = indices.contiguous();
  at::Tensor contiguous_weight = weight.contiguous();

  std::vector<int64_t> output_shape;
  output_shape.insert(output_shape.end(), indices.sizes().begin(), indices.sizes().end());
  output_shape.push_back(N);
  at::Tensor output =
      at::empty(output_shape, at::TensorOptions().dtype(weight.dtype()).device(indices.device()));
  const TritonJITFunction &f1 =
      TritonJITFunction::get_instance(std::string(utils::get_flag_gems_src_path() / "ops" / "embedding.py"),
                                      "embedding_kernel");
  c10::DeviceGuard guard(output.device());
  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
  CUstream raw_stream = static_cast<CUstream>(stream.stream());
  /*
  def embedding_kernel(
      out_ptr,  # pointer to the output
      in_ptr,  # pointer to the input
      weight_ptr,  # pointer to the weights
      N: tl.constexpr,  # number of columns in X
      BLOCK_SIZE: tl.constexpr,
  ):*/
  f1(raw_stream,
     M,
     1,
     1,
     /* num_warps = */ 8,
     /* num_stages = */ 1,
     output,
     contiguous_indices,
     contiguous_weight,
     N,
     BLOCK_SIZE);

  return output;
}

at::Tensor embedding_backward(const at::Tensor &grad_outputs,
                              const at::Tensor &indices,
                              int64_t num_weights,
                              int64_t padding_idx,
                              bool scale_grad_by_freq,
                              bool sparse) {
  TORCH_CHECK(!sparse, "Currently do not support sparse format");
  int64_t M = indices.numel();
  int64_t N = grad_outputs.size(-1);
  at::Tensor grad_inputs;
  if (grad_outputs.dtype() == at::kBFloat16) {
    grad_inputs = at::zeros({num_weights, N}, grad_outputs.options().dtype(at::kFloat));
  } else {
    grad_inputs = at::zeros({num_weights, N}, grad_outputs.options());
  }
  at::Tensor indice_freq;
  if (scale_grad_by_freq) {
    indice_freq = at::zeros({num_weights}, indices.options().dtype(at::kInt));
    int64_t INDICE_BLOCK_SIZE = 256;
    int64_t indice_grid = (M + INDICE_BLOCK_SIZE - 1) / INDICE_BLOCK_SIZE;
    const TritonJITFunction &f2 =
        TritonJITFunction::get_instance(std::string(utils::get_flag_gems_src_path() / "ops" / "embedding.py"),
                                        "indice_freq_kernel");
    c10::DeviceGuard guard(grad_outputs.device());
    c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
    CUstream raw_stream = static_cast<CUstream>(stream.stream());
    /*
    def indice_freq_kernel(
        indices_freq,
        indices,  # pointer to the input
        elem_cnt: tl.constexpr,  # number of columns in X
        INDICE_BLOCK_SIZE: tl.constexpr,
    ):*/
    f2(raw_stream,
       indice_grid,
       1,
       1,
       /* num_warps = */ 8,
       /* num_stages = */ 1,
       indice_freq,
       indices,
       M,
       INDICE_BLOCK_SIZE);
  }
  int64_t BLOCK_SIZE = utils::next_power_of_2(N);
  bool HAS_PADDING_IDX = (padding_idx != -1);
  const TritonJITFunction &f3 =
      TritonJITFunction::get_instance(std::string(utils::get_flag_gems_src_path() / "ops" / "embedding.py"),
                                      "embedding_backward_kernel");
  c10::DeviceGuard guard(grad_outputs.device());
  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
  CUstream raw_stream = static_cast<CUstream>(stream.stream());
  /*
  def embedding_backward_kernel(
      grad_in,  # pointer to the gradient input
      grad_out,  # pointer to the gradient output
      indices,  # pointer to the input
      padding_idx,  # padding_idx
      HAS_PADDING_IDX: tl.constexpr,
      N: tl.constexpr,  # number of columns in X
      BLOCK_SIZE: tl.constexpr,
  ):*/
  f3(raw_stream,
     M,
     1,
     1,
     /* num_warps = */ 8,
     /* num_stages = */ 1,
     grad_inputs,
     grad_outputs,
     indices,
     padding_idx,
     HAS_PADDING_IDX,
     N,
     BLOCK_SIZE);
  if (scale_grad_by_freq) {
    const TritonJITFunction &f4 =
        TritonJITFunction::get_instance(std::string(utils::get_flag_gems_src_path() / "ops" / "embedding.py"),
                                        "embedding_grad_scale_kernel");
    c10::DeviceGuard guard(grad_outputs.device());
    c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
    CUstream raw_stream = static_cast<CUstream>(stream.stream());
    /*
    def embedding_grad_scale_kernel(
        grad_out,
        indice_freq,
        n_rows,
        N,
        BLOCK_SIZE: tl.constexpr,
    ):*/
    f4(raw_stream,
       M,
       1,
       1,
       /* num_warps = */ 8,
       /* num_stages = */ 1,
       grad_inputs,
       indice_freq,
       num_weights,
       N,
       BLOCK_SIZE);
  }

  if (grad_outputs.dtype() == at::kBFloat16) {
    return grad_inputs.to(at::kBFloat16);
  }
  return grad_inputs;
}
}  // namespace flag_gems
